### PR TITLE
Jitclass Infer Spec from Type Annotations

### PR DESCRIPTION
**Pull request by EPronovost**
Extend a jitclass spec from type annotations on the class.
To enable this, we cleaned up `typeof` and extended it to support `typing` generics.  We also made the `jitclass` function more flexible for various use cases.  We can now do things like

```
@jitclass
class Counter:
    x: int
    
    def __init__(self):
        self.x = 0

    @property
    def count(self) -> int:
        return self.x

    def inc(self) -> int:
        self.x += 1
        return self.x
```

Related to issues https://github.com/numba/numba/issues/5218 and https://github.com/numba/numba/issues/2947.

### CODE REVIEW

File: examples/jitclass.py
Comment (StanczakDominik) [inline examples/jitclass.py:16]: I wonder if we shouldn't have both ways of doing things in this example, with type annotations and the old style spec? I do like the new style, though, so if the old one were to be phased out, I would shed no tears.

**Comment by EPronovost** [comment]
Hi @stuartarchibald , I saw this got moved to milestone 0.51.  Is there anything I can do, or will we just need to wait a few more months on this?  I know from the meeting minutes a while ago there was interest in this PR and a note that it may require discussion.  When should we expect that discussion to happen?

**Comment by stuartarchibald** [comment]
> Hi @stuartarchibald , I saw this got moved to milestone 0.51.

Yes, sorry, the view was that this is sufficiently involved and hasn't been discussed/explored enough.

> Is there anything I can do, or will we just need to wait a few more months on this?

For 0.50 we're trialing this: https://groups.google.com/a/continuum.io/forum/#!topic/numba-users/DUJhuX8uD_I helping out would free up more time to deal with involved features, though I obviously can't guarantee anything!

> I know from the meeting minutes a while ago there was interest in this PR and a note that it may require discussion. When should we expect that discussion to happen?

All PRs are discussed first in weekly triage, and then those which are proposing significant change are discussed at core developer meetings. I can propose this for discussion again at the next meeting to prompt others to form views in the next few weeks. Discussion of new features/changes often takes multiple iterations and unfortunately there's a bit of a backlog, for example the proposed https://github.com/numba/numba/pull/5560 and https://github.com/numba/numba/pull/5579 have been through multiple iterations of discussion.

File: numba/tests/test_typeof.py
Comment (eric-wieser) [inline numba/tests/test_typeof.py:315]: It's not clear to me that `typeof(x) == typeof(type(x))` makes any sense.
I'd expect these to return `numba.types.TypeRef`, if not the `ValueError` they give today.

Would it make sense to introduce a new `as_numba_type` function?

File: docs/source/user/jitclass.rst
Comment (EPronovost) [reply docs/source/user/jitclass.rst:82]: I like that better as well.  Done.

File: examples/jitclass.py
Comment (EPronovost) [reply examples/jitclass.py:16]: Done.

File: numba/tests/test_typeof.py
Comment (eric-wieser) [reply numba/tests/test_typeof.py:315]: In fact, this change is incompatible with the behavior today:
```
In [22]: numba.typeof(np.uint8(1))
Out[22]: uint8

In [23]: numba.typeof(type(np.uint8(1)))
Out[23]: class(uint8)
```

Comment (EPronovost) [reply numba/tests/test_typeof.py]: I had a reason for making it part of `typeof` originally, but can't remember it.  Making a new `as_numba_type` gives a cleaner API and offers some cool opportunities for user extensions.

**Comment by EPronovost** [comment]
Per @eric-wieser 's suggestion, I moved the type inference logic to a new `as_numba_type`.  This allowed me to make a new registry, which has interesting opportunities for user extensions (e.g. for other `jitclass`es or to modify the default behavior for `float`).  Currently there's some similar code between `typeof` and `as_numba_type`.  We could possibly do something like 
```
ty = as_numba_type(type(val))
if ty is not None:
    return ty
```
to replace all of the `typeof_impl` implementations of the form
```
@typeof_impl.register(py_type)
def impl(_, _):
    return nb_type
```

**Comment by esc** [comment]
@EPronovost thanks for updating this, I have labeled it as "waiting on reviewer input now".

File: docs/source/extending/interval-example.rst
Comment (eric-wieser) [inline docs/source/extending/interval-example.rst:88]: This would be simpler as `Interval._numba_type_ = interval_type`, which already informs `typeof` today.

However, this is only really possible in contrived cases. In all the extension types I've used so far, I've needed something like `IntervalType(self.dtype)` instead of just a global `IntervalType()`.

The mapping between python types and numba types is frequently one-to-many - so while `as_numba_type` is useful for the trivial cases you have being used here, it doesn't scale to all use-cases.

Comment (EPronovost) [reply docs/source/extending/interval-example.rst:88]: I agree that the current way Numba requires explicit types up-front does not address all possible use cases in Python.  The `as_numba_type` registry is meant to work statically on type annotations at compile type, not dynamically on instances at runtime.  I believe the combination of a [generic type annotation](https://docs.python.org/3/library/typing.html#generics) and the `@as_numba_type.register` decorator does allow for something like the `LoggedVar` example at the link above.  That being said, I agree that there's still a lot of room to improve the flexibility of `jitclass`.  There are several open issues about so-called "generic" `jitclass`es.  That was something I was interested in once this PR is concluded.

**Comment by EPronovost** [comment]
Cross-posting this follow-on PR that uses the `as_numba_type` registry to implement a typed version of named tuples. https://github.com/numba/numba/pull/5840

**Comment by esc** [comment]
hey, everyone, just checking on the status of this one as it is part of the 0.51.0 milestone. Where are we at?

**Comment by EPronovost** [comment]
@esc I'm not clear what the review process needs to include.  Is there something more I need to do?

**Comment by esc** [comment]
@EPronovost I just had another look, I believe this is awaiting a review from the core-dev team.

File: numba/experimental/jitclass/decorators.py
Comment (sklam) [inline numba/experimental/jitclass/decorators.py:20]: isn't it `as_numba_type(T)` not `typeof(T)`?

File: numba/experimental/jitclass/base.py
Comment (sklam) [inline numba/experimental/jitclass/base.py:177]: Should we consider all `__annotations__` from the `__mro__`?

Consider this:

```python

In [8]: class A:
   ...:     x : int
   ...:

In [9]: A.__annotations__
Out[9]: {'x': int}

In [10]: class B(A):
    ...:     pass
    ...:

In [11]: B.__annotations__   # B didn't add new __annotations__ and carries A.__annotations__
Out[11]: {'x': int}

In [12]: class C(A):
    ...:     y : int
    ...:

In [13]: C.__annotations__   # C has new __annotations__ which should be consider incremental
Out[13]: {'y': int}
```

File: docs/source/user/jitclass.rst
Comment (sklam) [inline docs/source/user/jitclass.rst:44]: The heading should also mention "type annotations", which is a more familiar term to look for.

File: docs/source/user/jitclass.rst
Comment (sklam) [inline docs/source/user/jitclass.rst:57]: Perhaps mention that type annotations on method arguments and return type are ignored since they are used in the example.

File: docs/source/extending/interval-example.rst
Comment (sklam) [reply docs/source/extending/interval-example.rst:88]: The comment:

> The as_numba_type registry is meant to work statically on type annotations at compile type, not dynamically on instances at runtime. 

needs to be documented.

File: docs/source/user/jitclass.rst
Comment (EPronovost) [reply docs/source/user/jitclass.rst:72]: > it's possible to make having type annotations bans the use of the spec

I'm not sure if this would work for numpy types.  Any array would get a type annotation `np.ndarray`, but numba requires more information.  In a future where numpy has generic type annotations for array this could be solved, but I don't think we're there yet.  This situation is why I thought to allow overwriting annotations with the `spec`.

If you remove the `@jitclass` decorator from the above example, that's still not valid Python code (also the type annotations for `Counter` seem incorrect).  What are you suggesting to change because of this example?  Add a check that every field is initialized in `__init__`?

File: numba/__init__.py
Comment (EPronovost) [reply numba/__init__.py:21]: Good point.  If we move it to `numba.experimental`, could we also just not create a deprecated endpoint with a warning in `numba/__init__.py`?

File: docs/source/user/jitclass.rst
Comment (EPronovost) [reply docs/source/user/jitclass.rst]: Good call.  Done.

File: docs/source/user/jitclass.rst
Comment (EPronovost) [reply docs/source/user/jitclass.rst]: Done.

File: numba/experimental/jitclass/base.py
Comment (EPronovost) [reply numba/experimental/jitclass/base.py]: Yup! TIL that `__annotations__` doesn't include fields from parent classes.  Updated and added a test for this.

File: numba/experimental/jitclass/decorators.py
Comment (EPronovost) [reply numba/experimental/jitclass/decorators.py]: Yup. Updated.

File: numba/__init__.py
Comment (EPronovost) [reply numba/__init__.py]: Removed from the top-level namespace.

File: docs/source/extending/interval-example.rst
Comment (EPronovost) [reply docs/source/extending/interval-example.rst:84]: For sure.  I added a section to the `types` doc as well.

File: docs/source/extending/interval-example.rst
Comment (EPronovost) [reply docs/source/extending/interval-example.rst:88]: Done.

**Comment by EPronovost** [comment]
@sklam thanks for the review! I've updated the PR.

File: docs/source/user/jitclass.rst
Comment (sklam) [reply docs/source/user/jitclass.rst]: You convinced me. We need to have both spec and annotations. 

> Add a check that every field is initialized in __init__?

We'll need to address the segfault in a different PR. Prevent the segfault or check that all fields are initialized, or both.

Comment (EPronovost) [reply docs/source/user/jitclass.rst]: Ok so is there any action item for this comment thread (beyond the documentation updates)?

File: docs/source/reference/types.rst
Comment (eric-wieser) [inline docs/source/reference/types.rst:359]: Perhaps worth allowing (or requiring) the spelling
```suggestion
      ...     x: typing.Annotated[int, numba.int64]
```

File: docs/source/user/jitclass.rst
Comment (eric-wieser) [inline docs/source/user/jitclass.rst:65]: Again, this could be
```suggestion
        y: Annotated[np.ndarray, float64[:]]
```

File: docs/source/reference/types.rst
Comment (EPronovost) [reply docs/source/reference/types.rst:362]: I didn't realize PEP 593 had been back-ported in `typing_extensions` already.  Awesome!  Added support for it.  See the updated docs and examples.

File: docs/source/user/jitclass.rst
Comment (EPronovost) [reply docs/source/user/jitclass.rst:65]: Yup! See comment above

File: docs/source/reference/types.rst
Comment (eric-wieser) [inline docs/source/reference/types.rst:383]: I don't think this is a good idea - why not just require the user to use the normal syntax for array types rather than inventing a new one?

**Comment by EPronovost** [comment]
@eric-wieser suggested supporting `typing_extensions.Annotated`, which I think is a great idea (I didn't realize it was in `typing_extensions` already).  This requires adding the `typing_extensions` package as a dependency, which I've already done in https://github.com/numba/numba/pull/6222.  My hope would be to merge that PR (which should hopefully be quick to review) and then update and merge this one.  I tested locally under Python 3.6, 7, and 8 (since a lot of `typing` changed in 3.7).

File: docs/source/reference/types.rst
Comment (EPronovost) [reply docs/source/reference/types.rst:383]: With these options, the user doesn't need to use any numba types for some cases.  For instance you can do
```
Annotated[np.ndarray, float, 2]
```
I think this lowers the barrier for entry with new users... they don't even have to read anything about numba types and what `__getitem__` does.  By just using `Annotated[np.ndarray, {float, int, complex, bool}, N]` they can get pretty far.

File: numba/core/typing/asnumbatype.py
Comment (eric-wieser) [inline numba/core/typing/asnumbatype.py:80]: The spec for annotated says you should ignore any arguments you don't recognize, as they may be for other tools. So I'd recommend you filter the annotations to those where `ininstance(arg, numba.Type)` is true, and assert that is of length 1.

Comment (EPronovost) [reply numba/core/typing/asnumbatype.py:80]: The [PEP](https://www.python.org/dev/peps/pep-0593/#consuming-annotations) also says
> Multiple annotations: It's up to the tool consuming the annotations to decide whether the client is allowed to have several annotations on one type and how to merge those annotations.

Checking `isinstance(arg, numba.Type)` would disallow the current recursive call where `Annotated[py_type, x] => as_numba_type(x)`.

I'll leave it up to the core devs.  Either we only support explicit numba types, in which case we can do such a check, or we also support numpy and builtin types (e.g. `Annotated[np.ndarray, int, 1]`), in which case I think it's simpler to say that our tool does not support multiple annotations. 

FWIW I wouldn't expect numba to "just work" with other python tools, so supporting their annotations may be misleading.  I don't have much experience with what other tools would use `Annotated` for.

File: docs/source/reference/types.rst
Comment (eric-wieser) [reply docs/source/reference/types.rst:383]: I don't think it's numba's place to advocate for a way to statically type numpy arrays. There are ongoing discussions in numpy about how to annotate arrays with shape and dtype information, inventing a new and different protocol to whatever is standardised there will IMO just cause confusion. `Annotated[np.ndarray, float, 2]` is barely shorter than `Annotated[np.ndarray, numba.Array(float, 2)]`, and the later makes it very clear to the reader that the syntax is numba-specific.

Comment (sklam) [reply docs/source/reference/types.rst:383]: I'd also prefer to not invent new ways to annotate ndarrays. `Annotated[np.ndarray, numba.Array(float, 2)]` looks more understandable than `Annotated[np.ndarray, float, 2]`.

File: numba/core/typing/asnumbatype.py
Comment (sklam) [reply numba/core/typing/asnumbatype.py:80]: I don't think I am familiar with the type annotation standard or convention to make the call.

**Comment by sklam** [comment]
@EPronovost, we are near our 0.52.0rc due date. Besides the merge conflict, https://github.com/numba/numba/pull/5609/commits/7a8c080a715e62333922c3b966da8c0649a8c57e seems to be controversial and I am not well-versed in type annotation to make any decision. At the same time, I don't want this PR to drag out too long. Is it possible to revert that commit and merge what we have knowing that this feature (along with jitclass) is experimental.

**Review by sklam — APPROVED** [review]
Thanks for the contribution. It's looking good.

**Pull request merged**
